engine/src/main/java/org/terasology/logic/location/LocationChangedEvent.java

```diff
@@ -0,0 +1,83 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.logic.location;
+
+import org.terasology.entitySystem.event.Event;
+import org.terasology.math.geom.BaseVector3f;
+import org.terasology.math.geom.ImmutableQuat4f;
+import org.terasology.math.geom.ImmutableVector3f;
+import org.terasology.math.geom.Quat4f;
+import org.terasology.math.geom.Vector3f;
+
+public class LocationChangedEvent implements Event {
+    public final LocationComponent component;
+    public final ImmutableVector3f oldPosition;
+    public final ImmutableQuat4f oldRotation;
+    public final ImmutableVector3f newPosition;
+    public final ImmutableQuat4f newRotation;
+
+    public LocationChangedEvent(LocationComponent newLocation) {
+        this(newLocation, newLocation.position, newLocation.rotation, newLocation.position, newLocation.rotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Vector3f oPosition) {
+        this(newLocation, oPosition, newLocation.rotation, newLocation.position, newLocation.rotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Vector3f oPosition, Vector3f nPosition) {
+        this(newLocation, oPosition, newLocation.rotation, nPosition, newLocation.rotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Quat4f oRotation) {
+        this(newLocation, newLocation.position, oRotation, newLocation.position, newLocation.rotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Quat4f oRotation, Quat4f nRotation) {
+        this(newLocation, newLocation.position, oRotation, newLocation.position, nRotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Vector3f oPosition, Quat4f oRotation) {
+        this(newLocation, oPosition, oRotation, newLocation.position, newLocation.rotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Vector3f oPosition, Quat4f oRotation, Vector3f nPosition) {
+        this(newLocation, oPosition, oRotation, nPosition, newLocation.rotation);
+    }
+
+    public LocationChangedEvent(LocationComponent newLocation, Vector3f oPosition, Quat4f oRotation, Quat4f nRotation)
+    {
+        this(newLocation, oPosition, oRotation, newLocation.position, nRotation);
+    }
+
+    public LocationChangedEvent(LocationComponent nComponent, Vector3f oPosition, Quat4f oRotation, Vector3f nPosition, Quat4f nRotation)
+    {
+        oldPosition = ImmutableVector3f.createOrUse(oPosition);
```

Comment:
This would always create a new `ImmutableVector3f` because `Vector3f` and it's immutable counterpart are not related. So every time a new `LocationChangedEvent` is created, these 4 new objects will be created again.

![image](https://user-images.githubusercontent.com/32035777/69480021-c0d73a00-0e29-11ea-850a-523116400c72.png)

Reply:
The point is that it is *not* for blocks -- it is meant to be block *item* entities, players, animals, etc. This allows the existence of sensors that check for the presence of such mobile entities without requiring a collision check, or for each sensor to constantly scan all entities for their location.

I will check again and see if immutable is unnecessary, but my recollection is that it is required so that event recipients cannot modify the very mutable "last position" vector and mess up any other event recipients.